src/uvc_ros_driver.cpp

```diff
@@ -40,1372 +40,899 @@
-	// uvc_device_descriptor_t *desc;
-	// uvc_get_device_descriptor(dev_, &desc);
-	// uvc_free_device_descriptor(desc);
-
-	/* Try to negotiate a 640x480 30 fps YUYV stream profile */
-	res = uvc_get_stream_ctrl_format_size(
-		      devh_, &ctrl_,              /* result stored in ctrl */
-		      UVC_FRAME_FORMAT_YUYV,      /* YUV 422, aka YUV 4:2:2. try _COMPRESSED */
-		      raw_width_, raw_height_, 30 /* width, height, fps */
-	      );
-
-	if (res < 0) {
-		/* device doesn't provide a matching stream */
-		uvc_perror(res, "get_mode");
-		ROS_ERROR("Device doesn't provide a matching stream");
-		return res;
-	}
-
-	return res;
+uvc_error_t uvcROSDriver::initAndOpenUvc() {
+  uvc_error_t res;
+  /* Initialize a UVC service context. Libuvc will set up its own libusb
+  * context. Replace NULL with a libusb_context pointer to run libuvc
+  * from an existing libusb context. */
+  res = uvc_init(&ctx_, NULL);
+
+  if (res < 0) {
+    uvc_perror(res, "uvc_init");
+    ROS_ERROR("Unable to initialize uvc service context");
+    return res;
+  }
+
+  /* Locates the first attached UVC device, stores in dev */
+  /* filter devices: vendor_id, product_id, "serial_num" */
+  res = uvc_find_device(ctx_, &dev_, 0x04b4, 0, NULL);
+
+  if (res < 0) {
+    uvc_perror(res, "uvc_find_device"); /* no devices found */
+    ROS_ERROR("No devices found");
+    return res;
+  }
+
+  ROS_INFO("Device found");
+
+  /* Try to open the device: requires exclusive access */
+  res = uvc_open(dev_, &devh_);
+
+  if (res < 0) {
+    uvc_perror(res, "uvc_open"); /* unable to open device */
+    ROS_ERROR("Unable to open the device");
+    return res;
+  }
+
+  /* Try to negotiate a 640x480 30 fps YUYV stream profile */
+  res = uvc_get_stream_ctrl_format_size(
+      devh_, &ctrl_,              /* result stored in ctrl */
+      UVC_FRAME_FORMAT_YUYV,      /* YUV 422, aka YUV 4:2:2. try _COMPRESSED */
+      raw_width_, raw_height_, 30 /* width, height, fps */
+      );
+
+  if (res < 0) {
+    /* device doesn't provide a matching stream */
+    uvc_perror(res, "get_mode");
+    ROS_ERROR("Device doesn't provide a matching stream");
+    return res;
+  }
+
+  return res;
 }
 
-////////////////////////////////////////////////////////////////////////////////
+uint8_t *uvcROSDriver::rawDataPtr(const uvc_frame_t *frame, const size_t line,
+                                  const size_t offset) {
+  return &(static_cast<uint8_t *>(
+      frame->data)[2 * ((line + 1) * frame->width - offset)]);
+}
 
-int16_t uvcROSDriver::ShortSwap(int16_t s)
-{
-	unsigned char b1, b2;
-	b1 = s & 255;
-	b2 = (s >> 8) & 255;
-	return (b1 << 8) + b2;
+uint8_t uvcROSDriver::readUInt8(const uvc_frame_t *frame, const size_t line,
+                                const size_t offset) {
+  return *rawDataPtr(frame, line, offset);
 }
 
-////////////////////////////////////////////////////////////////////////////////
+int16_t uvcROSDriver::readInt16(const uvc_frame_t *frame, const size_t line,
```

Comment:
Nice! I guess the person who wrote the driver was not aware of the concept of "function" and simply copied and past the same code to convert raw bytes into int* variables xD